src/lib/flash-helper.js

```diff
@@ -204,9 +177,56 @@ async function createFlashSteps({ modules, isInDfuMode, platformId }) {
 	}
 }
 
+
+async function flashDeviceInNormalMode(device, data, { name, progress } = {}) {
+	if (device.isInDfuMode) {
+		// put device in normal mode
+		if (progress) {
+			progress({ event: 'switch-mode', mode: 'normal' });
+		}
+		device = await usbUtils.reopenInNormalMode(device, { reset: true });
+	}
+	// put the device in listening mode to prevent cloud connection
+	try {
+		await device.enterListeningMode();
+	} catch (error) {
+		// ignore if the device is already in listening mode
+	}
+
+
+	// flash the file in normal mode
+	if (progress) {
+		progress({ event: 'flash-file', filename: name });
+	}
+	await device.updateFirmware(data, { progress, timeout: FLASH_TIMEOUT });
+
+	// wait for the device to apply the firmware
+	await delay(FLASH_APPLY_DELAY);
+	device = await usbUtils.reopenInNormalMode(device, { reset: false });
+	return device;
+}
+
+async function flashDeviceInDfuMode(device, data, { name, altSetting, startAddr, progress } = {}) {
+	if (!device.isInDfuMode) {
+		// put device in dfu mode
+		if (progress) {
+			progress({ event: 'switch-mode', mode: 'DFU' });
+		}
+		device = await usbUtils.reopenInDfuMode(device);
+	}
+
+	// flash the file over DFU
+	if (progress) {
+		progress({ event: 'flash-file', filename: name });
+	}
+	await device.writeOverDfu(data, { altSetting, startAddr: startAddr, progress });
+}
+
```

Comment:
I moved the functionality to put the devices into normal and dfu modes into separate functions, such that they can be used by `particle flash --usb` implementation.